🔨️refactor : 비로그인시 사용가능한 api 구조 변경

jickDo opened the pull request:
<!--
체크하려면 괄호 안에 "x"를 입력하세요.
각 규칙은 Convention 문서에 있습니다.
PR 제목에 쓰는 prefix는 다음과 같습니다.

✨ feat : 새로운 기능
🔨️ refactor : 코드 리팩토링
🐎 perf : 성능을 향상
🐛 fix : 버그를 고칠 때
🧪 test : 테스트 코드
🚜 rename : 파일 이름 변경 혹은 구조를 변경
🚀 deployment : 배포 / 개발 작업과 관련된 모든 것
🔥 remove : 코드 또는 파일 제거
📚 docs : 문서
📝 chore : 사소한 코드 또는 언어를 변경 기타 변경사항 (빌드 스크립트 수정, 패키지 매니징 설정 등)

-->

## PR 타입

<!-- 어떤 유형의 PR인지 체크해주세요. -->

<!-- 체크하려면 괄호 안에 "x"를 입력하세요. -->

- [ ] Feature
- [x] Refactoring (no functional changes, no api changes)
- [ ] Performance Improvement
- [ ] Bugfix
- [ ] Test Code
- [ ] Code style update (formatting, local variables)
- [ ] Other... Please describe:

## 개요

- 비로그인시 사용가능한 api에 대한 구조 변경

## 작업 및 변경 사항

- 비로그인 전용 인가 어노테이션 구현
- 리뷰 리스트 조회
- 리뷰 상세 조회
- 인기 여행 조회
- + 인기 여행 조회에서 북마크 여부를 추가함

### 공유 하고 싶은 내용 및 참고 자료

- 메인에 나오는 리뷰와 인기 여행지기 조회의 경우 따로 비로그인과 로그인 구분되는 값이 오는게 아니라 변경사항이 없습니다.
- 다만 , 여행조회에서 여행 북마크 여부와 여행 상세 조회에서 여행참여 여부를 반환해야 할것 같은데 이 부분 어떻게 생각하시나요?

++ 로그아웃 기능구현

close #102

jickDo committed:
[chore] submodule

jickDo committed:
[feat] 비로그인 전용 인가 어노테이션 생성

jickDo committed:
[refactor] 리뷰 리스트 페이지네이션 비로그인 구현

jickDo committed:
[refactor] 리뷰 자세히 보기 비로그인 구현

jickDo committed:
[refactor] 이번주 인기여행 조회 북마크 추가 및 비로그인 구현

hyunmuam commented on src/main/java/potatoes/server/repository/ReviewRepository.java:
```diff
@@ -22,11 +22,12 @@ public interface ReviewRepository extends JpaRepository<Review, Long> {
-		        THEN TRUE 
-		        ELSE FALSE 
-		    END,
+		    (CASE
+		        WHEN :userId = -1 THEN CAST(NULL AS boolean)
+		        WHEN EXISTS (SELECT 1 FROM ReviewLike rl WHERE rl.review = r AND rl.user.id = :userId)
+		        THEN true
+		        ELSE false
+		    END),
```
P4: `userId`가 -1일 때 CASE문을 사용하는 것보다 별도로 분리하는 것이 성능상 더 좋을 것 같습니다. 분리함으로써 데이터베이스는 각 쿼리를 별도로 최적화할 수 있으며, 모든 행에 대해 CASE문을 평가하는 오버헤드를 피할 수 있지 않을까 생각합니다.

jickDo replied:
비로그인 확인여부는 비즈니스 로직에서 검사하는 방향으로 바꿔보겠습니다.

jickDo committed:
[refactor] api 위치 재정리

jickDo committed:
[refactor!] 리뷰 리스트 조회 변경

리뷰 리스트 조회시 기존 비로그인 검사를 jpa영역에서 진행하며 불필요한 case구문을 매번 실행하는 로직에서 비로그인 검사 케이스를 비즈니스 로직으로 가져오고 오버로딩을 통해진행

jickDo committed:
[refactor] 여행 컨트롤러 메서드 위치 조정

jickDo committed:
[refactor] 로그아웃 구현

hyunmuam reviewed (approved):
P4: "여행조회에서 여행 북마크 여부와 여행 상세 조회에서 여행참여 여부를 반환해야 할것 같은데 이 부분 어떻게 생각하시나요?" 
- 음... 고민이네요. `UserId`를 -1이 아니라 옵셔널로 가져오는게 더 편하게 작업할 수 있지 않을까 생각하긴합니다. 이 이슈&브랜치에서 계속 작업하셔도되고 새로운 이슈로 분기하셔도 될 것 같아요,

jickDo commented:
> P4: "여행조회에서 여행 북마크 여부와 여행 상세 조회에서 여행참여 여부를 반환해야 할것 같은데 이 부분 어떻게 생각하시나요?"
> 
> * 음... 고민이네요. `UserId`를 -1이 아니라 옵셔널로 가져오는게 더 편하게 작업할 수 있지 않을까 생각하긴합니다. 이 이슈&브랜치에서 계속 작업하셔도되고 새로운 이슈로 분기하셔도 될 것 같아요,

이 부분은 생각못해봤는데 확실히 optional로 풀어내는게 좋아보이긴 하네요

남은 비로그인 대응 관련 api 두개가 여행조회, 여행 상세 조회 부분인데 이부분에서 각각 여행 북마크 여부와 여행 참여 여부를 더 보내줘야 하다보니 비로그인 대응 이슈를 넘어서 api구조 변경이 일어나게 됩니다. 그렇기 때문에 이슈를 분리하는게 좋아보이고 분리한다면 이 이슈 이후에 여행조회, 여행상세 조회 이슈 지금처럼 하고 기존 -1처리 로직을 전부 optional로 리팩토링 해보는 이슈로 빼는게 좋아보여요

요약: 현재 이슈 머지 -> 여행, 여행 상세 조회 반환값 변경및 비로그인 구현 -> 전체 비로그인 -1처리여부 optional로 변경 요렇게 가고자합니다.

jickDo committed:
Merge remote-tracking branch 'origin/develop' into refactor/#102
